library/init/meta/widget/basic.lean

```diff
@@ -172,10 +172,19 @@ inductive mouse_event_kind
 | on_mouse_enter
 | on_mouse_leave
 
-/-- An effect is an action at the root of the widget component hierarchy
-and can give instructions to the editor to perform some task. -/
+/-- An effect is some change that the widget makes outside of its own state.
+Usually, giving instructions to the editor to perform some task.
+- `insert_text_relative` will insert at a line relative to the position of the widget.
+- `insert_text_absolute` will insert text at the precise position given.
+- `reveal_position` will move the editor to view the given position.
+- `highlight_position` will add a text highlight to the given position.
+- `clear_highlighting` will remove all highlights created with `highlight_position`.
+- `copy_text` will copy the given text to the clipboard.
+- `custom` can be used to pass custom effects to the client without having to recompile Lean.
+-/
 meta inductive effect : Type
-| insert_text (text : string)
+| insert_text_absolute (file_name : option string) (p : pos) (text : string)
+| insert_text_relative (relative_line : int) (text : string)
```

Comment:
Does it make sense to include `relative_col` too?

Reply:
AFAIU it inserts a whole line, with matching indentation.  So column wouldn't make much sense.